src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs

```diff
@@ -8666,5 +8666,331 @@ static async System.Threading.Tasks.Task Main()
+  IL_0149:  ldc.i4.0
+  IL_014a:  call       "void System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<bool>.SetResult(bool)"
+  IL_014f:  ret
+  IL_0150:  ldarg.0
+  IL_0151:  ldflda     "System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<bool> C.<M>d__0.<>v__promiseOfValueOrEnd"
+  IL_0156:  ldc.i4.1
+  IL_0157:  call       "void System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<bool>.SetResult(bool)"
+  IL_015c:  ret
+}
+"""");
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/74013")]
+        public void ClearCurrentWhenAwaiting_ManagedType()
+        {
+            var src = """
+using System.Threading.Tasks;
+using System.Collections.Generic;
+
+var tcs = new TaskCompletionSource();
+var enumerable = C.M(tcs.Task);
+var enumerator = enumerable.GetAsyncEnumerator();
+if (!await enumerator.MoveNextAsync())
+    throw null;
+
+System.Console.Write(enumerator.Current.field);
+
+var promise = enumerator.MoveNextAsync();
+System.Console.Write(enumerator.Current.field is null);
+
+tcs.SetResult();
+
+if (!await promise)
+    throw null;
+
+System.Console.Write(enumerator.Current.field);
+
+public struct S
+{
+    public object field;
+}
+
+public class C
+{
+    public static async IAsyncEnumerable<S> M(Task t)
+    {
+        object o = "first ";
+        yield return new S { field = o };
+        await t;
+        yield return new S { field = " second" };
+    }
+}
+""";
+            var comp = CreateCompilation(src, targetFramework: TargetFramework.Net80);
+            var verifier = CompileAndVerify(comp,
+                expectedOutput: ExecutionConditionUtil.IsMonoOrCoreClr ? "first True second" : null,
+                verify: ExecutionConditionUtil.IsMonoOrCoreClr ? Verification.Passes : Verification.Skipped).VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/74013")]
+        public void ClearCurrentWhenAwaiting_ValueType()
+        {
+            var src = """
+using System.Threading.Tasks;
+using System.Collections.Generic;
+
+var tcs = new TaskCompletionSource();
+var enumerable = C.M(tcs.Task);
+var enumerator = enumerable.GetAsyncEnumerator();
+if (!await enumerator.MoveNextAsync())
+    throw null;
+
+System.Console.Write(enumerator.Current);
+
+var promise = enumerator.MoveNextAsync();
+System.Console.Write(enumerator.Current);
+
+tcs.SetResult();
+
+if (!await promise)
+    throw null;
+
+System.Console.Write(enumerator.Current);
+
+public class C
+{
+    public static async IAsyncEnumerable<int> M(Task t)
+    {
+        yield return 42;
+        await t;
+        yield return 43;
+    }
+}
+""";
+            var comp = CreateCompilation(src, targetFramework: TargetFramework.Net80);
+            var verifier = CompileAndVerify(comp,
+                expectedOutput: ExecutionConditionUtil.IsMonoOrCoreClr ? "424243" : null,
```

Comment:
>424243 [](http://example.com/codeflow?start=74&length=6)

It feels to me that it would be better to invalidate the value of Current, or, perhaps, even throw InvalidOperationException from that property during the period of time starting with MoveNextAsync invocation and until the return task completes. And do that regardless of whether the element type is managed or not.

Reply:
That feels like addressing a different problem. The problem we're trying to solve here is not callers accessing the value of `Current` after the state machine has restarted, but simply that we're preventing GC of referenced values.

@stephentoub What do you think of clearing the `Current` value regardless of whether the element type is managed or not?

Reply:
> but simply that we're preventing GC of referenced values

Thank you for making the primary motivation clear. However with the current approach we end up with an observable inconsistent behavior. Users are likely to build up a "muscle memory" with unmanaged types thinking that values are available, and then run into a problem with managed types while doing the same thing.

Reply:
> What do you think of clearing the Current value regardless of whether the element type is managed or not?

I'm not particularly concerned with "consistency" here. The difference here is only observable when accessing Current at a time when you're not supposed to access Current, such that the behavior is undefined. We have a multitude of places where we have similar discrepancies, where we check RuntimeHelpers.IsReferenceOrContainsReference and clear based on it, e.g. only clearing arrays returned to the array pool if they contain a reference, or only zero'ing the data during `List<T>.Clear` if it contains a reference. Those are also similarly observable but only in cases where you're relying on undefined behavior / unsafe APIs, e.g. subsequently using CollectionsMarshal.SetCount to increase back the list's count, at which point it'll still contain the data from before the Clear but only if the data didn't contain references. And even if this does just always clear Current, there are still inconsistencies. For example, if the clearing happens as part of an await, and in particular as part of a suspension, then given the code:
```C#
yield return 42;
await t;
```
whether e.Current is 42 after the enumerator completes will be based on a race condition of whether `t` was completed by the time its IsCompleted was checked as part of the await.

Thus, I'd have no qualms if the zero'ing was only done for managed Ts. That said, in all but the most extreme cases, I also doubt there will be a measurable impact to the zero'ing, given that in the most common cases you're dealing with either a single reference or a small struct, such that zero'ing is just one or a few instructions, and it's only at a time when we're about to pay non-trivial costs as part of suspension.

Reply:
Jared pointed out that we have some precedent for clearing variables to avoid unnecessary reference retention: locals in nested blocks in async/iterator state machines.
See `MethodToStateMachineRewriter.AddVariableCleanup`.
So I think it's okay to keep the logic conditional for clearing `Current` as well.

Note: the logic `MightContainReferences` could be improved, but I'm planning to address that in a follow-up PR.

[sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQAYACdbAFgG4CD0BmMzEgYRIIG8CSezHz6AHlgAXAHwkAsgApgAezkAbEsACU7fLxJdNW3gHcAFlEUBTErNXc9PHTZvlSAZxIBeEgCIAbhEUBXUw8aXXtecmwyAHYSCkxg0K0WORgnJVMAOgB1ACcoEVNpJ1V4hJ4AejKSfXMAYzMIbJIAAycmkigwAAdFKBq8xQBPaxsAX2G9JJS0rNz86Q9sD2LxsfwRoA)
```
using System;
using System.Collections.Generic;

public class C 
{
    public IEnumerable<int> M(bool b) 
    {
        while (b)
        {
            string s = "value";
            yield return 42;
            Console.Write(s);
            // we clear `s` implicitly
        }
        Console.Write("1");
    }
}
```
[sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQAYACdbAVgG4CD0BmMzEgYRIIG8CSezH0AHGQBsAHlgAXAHwkAsgApgAeyUAbEsACU7fLxJdde3gHcAFlFUBTEos3cjPAw4flSAZxIBeEgCIAbhCqAK6WPjSGzrzoAJwiAHQAmlCWqgjymuGReixKMG5qlnEA6gBOUBKW8m4Z9lkkAPT1JMbWAMZWECUkAAZu3SRQYAAOqlCt5aoAnrVGAL4z2bn5VsVlFfI+2D41EUboAOwkACyYmbzz+LNAA==)
```
using System;
using System.Threading.Tasks;

public class C 
{
    public async Task<int> M(bool b) 
    {
        while (b)
        {
            string s = "value";
            await Task.Yield();
            Console.Write(s);
            // we clear `s` implicitly
        }
        Console.Write("1");
        return 42;
    }
}
```